prac_07/project_management.py

```diff
@@ -0,0 +1,216 @@
+"""
+Project management program.
+Estimated: 5 hours
+Actual: 5hr 19min
+"""
+from operator import attrgetter, itemgetter
+import datetime
+from prac_07.project import Project
+
+DEFAULT_FILENAME = "projects.txt"
+MENU = "- (L)oad projects\n- (S)ave projects\n- (D)isplay projects\n- (F)ilter projects by date\n- (A)dd new project\n- (U)pdate project\n- (Q)uit"
+
+
+def main():
+    """Load file then display menu. Will save file after quiting"""
+    print("Welcome to Pythonic Project Management")
+    projects = load_projects(DEFAULT_FILENAME)
+    print(f"Loaded {len(projects)} from {DEFAULT_FILENAME}")
+    print(MENU)
+    choice = input(">>> ").upper()
+    while choice != "Q":
+        if choice == "L":
+            projects = load_new_file()
+        elif choice == "S":
+            save_file(projects)
+        elif choice == "D":
+            display_project_completion_status(projects)
+        elif choice == "F":
+            display_projects_by_date(projects)
+        elif choice == "A":
+            add_new_project(projects)
+        elif choice == "U":
+            update_project_with_error_checking(projects)
+        else:
+            print("Invalid menu choice")
+        print(MENU)
+        choice = input(">>> ").upper()
+    save_file(projects)
+    print("Thank you for using custom-built project management software.")
+
+
+def display_projects_by_date(projects):
+    """Display projects after a certain date. Should sort by date."""
+    # date_filter = convert_to_datetime_object("20/7/2022")
+    date_filter = get_valid_date("Show projects that start after date (dd/mm/yyyy): ")
+    filtered_projects = [project for project in projects if
+                         convert_to_datetime_object(project.start_date) >= date_filter]
+    for project in sorted(filtered_projects, key=attrgetter("start_date")):
+        print(project)
+
+
+def update_project_with_error_checking(projects):
+    """Update chosen project with error checking."""
+    for project_number, project in enumerate(projects):
+        print(f"{project_number} {project}")
+    project_choice = int(input("Project choice: "))
+    while project_choice > len(projects) - 1:
+        print("Invalid project")
+        project_choice = int(input("Project choice: "))
+    print(f"Record {projects[project_choice]}")
+    new_percentage = input("New percentage: ")
```

Comment:
I believe this and the following if operator should have been independant functions, to follow DRY and modular code principles